File: core/src/main/java/info/nightscout/androidaps/plugins/pump/common/utils/ProfileUtil.java
package info.nightscout.androidaps.plugins.pump.common.utils;

import java.util.Locale;

import info.nightscout.androidaps.data.Profile;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType;

public class ProfileUtil {


    public static String getProfileDisplayable(Profile profile, PumpType pumpType) {

        StringBuilder stringBuilder = new StringBuilder();

        for (Profile.ProfileValue basalValue : profile.getBasalValues()) {

            double basalValueValue = pumpType.determineCorrectBasalSize(basalValue.value);

            int hour = basalValue.timeAsSeconds / (60 * 60);

            stringBuilder.append((hour < 10 ? "0" : "") + hour + ":00");

            stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", basalValueValue));
            stringBuilder.append(", ");
        }
        if (stringBuilder.length() > 3)
            return stringBuilder.substring(0, stringBuilder.length() - 2);
        else
            return stringBuilder.toString();
    }

    public static String getBasalProfilesDisplayable(Profile.ProfileValue[] profiles, PumpType pumpType) {

        StringBuilder stringBuilder = new StringBuilder();

        for (Profile.ProfileValue basalValue : profiles) {

            double basalValueValue = pumpType.determineCorrectBasalSize(basalValue.value);

            int hour = basalValue.timeAsSeconds / (60 * 60);

            stringBuilder.append((hour < 10 ? "0" : "") + hour + ":00");

            stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", basalValueValue));
            stringBuilder.append(", ");
        }
        if (stringBuilder.length() > 3)
            return stringBuilder.substring(0, stringBuilder.length() - 2);
        else
            return stringBuilder.toString();
    }


}

File: pump/pump-common/src/main/java/info/nightscout/pump/common/utils/ProfileUtil.kt
package info.nightscout.pump.common.utils

import info.nightscout.interfaces.profile.Profile
import info.nightscout.interfaces.profile.Profile.ProfileValue
import info.nightscout.interfaces.pump.defs.PumpType
import java.util.Locale

object ProfileUtil {

    fun getProfileDisplayable(profile: Profile, pumpType: PumpType): String {
        val stringBuilder = StringBuilder()
        for (basalValue in profile.getBasalValues()) {
            val basalValueValue = pumpType.determineCorrectBasalSize(basalValue.value)
            val hour = basalValue.timeAsSeconds / (60 * 60)
            stringBuilder.append((if (hour < 10) "0" else "") + hour + ":00")
            stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", basalValueValue))
            stringBuilder.append(", ")
        }
        return if (stringBuilder.length > 3) stringBuilder.substring(0, stringBuilder.length - 2) else stringBuilder.toString()
    }

    fun getBasalProfilesDisplayable(profiles: Array<ProfileValue>, pumpType: PumpType): String {
        val stringBuilder = StringBuilder()
        for (basalValue in profiles) {
            val basalValueValue = pumpType.determineCorrectBasalSize(basalValue.value)
            val hour = basalValue.timeAsSeconds / (60 * 60)
            stringBuilder.append((if (hour < 10) "0" else "") + hour + ":00")
            stringBuilder.append(" ")
            stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", basalValueValue))
            stringBuilder.append(",\n")
        }
        return if (stringBuilder.length > 3) stringBuilder.substring(0, stringBuilder.length - 2) else stringBuilder.toString()
    }

    fun getBasalProfilesDisplayableAsStringOfArray(profile: Profile, pumpType: PumpType): String {
        val stringBuilder = java.lang.StringBuilder()
        // for (basalValue in profiles) {
        //     val basalValueValue = pumpType.determineCorrectBasalSize(basalValue.value)
        //     val hour = basalValue.timeAsSeconds / (60 * 60)
        //     stringBuilder.append((if (hour < 10) "0" else "") + hour + ":00")
        //     stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", basalValueValue))
        //     stringBuilder.append(", ")
        // }
        // return if (stringBuilder.length > 3) stringBuilder.substring(0, stringBuilder.length - 2) else stringBuilder.toString()

        var entriesCopy = profile.getBasalValues()

        for (i in entriesCopy.indices) {
            val current = entriesCopy[i]
            // var currentTime = if (current.startTime_raw % 2 == 0) current.startTime_raw.toInt() else current.startTime_raw - 1
            // currentTime = currentTime * 30 / 60
            val currentTime = current.timeAsSeconds / (60 * 60)

            var lastHour: Int
            lastHour = if (i + 1 == entriesCopy.size) {
                24
            } else {
                val basalProfileEntry = entriesCopy[i + 1]
                //val rawTime = if (basalProfileEntry.startTime_raw % 2 == 0) basalProfileEntry.startTime_raw.toInt() else basalProfileEntry.startTime_raw - 1
                //rawTime * 30 / 60
                basalProfileEntry.timeAsSeconds / (60 * 60)
            }

            // System.out.println("Current time: " + currentTime + " Next Time: " + lastHour);
            for (j in currentTime until lastHour) {
                // if (pumpType == null)
                //     basalByHour[j] = current.rate
                // else
                //basalByHour[j] = pumpType.determineCorrectBasalSize(current.value)

                stringBuilder.append(String.format(Locale.ENGLISH, "%.3f", pumpType.determineCorrectBasalSize(current.value)))
                stringBuilder.append(" ")
            }
        }

        return stringBuilder.toString().trim()

    }

}